### PR TITLE
Feature 3.4:  Ability to hold and release new transactions

### PR DESCRIPTION
**Pull request by matthewvon**
This is a 3.4, rocksdb only change to TransactionManager.  Use existing ReadWriteLock class to pause the start of new transactions during rocksdb checkpoint (HotBackup marketing name) and restore.

The ReadWriteLock could be easily moved to the base class in the future to provide capability to mmfiles and such.

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBTransactionManager.h [hunk 1]
```diff
@@ -67,8 +69,17 @@ class RocksDBTransactionManager final : public TransactionManager {
 
   uint64_t getActiveTransactionCount() override { return _nrRunning; }
 
+  // temporarily block all new transactions
+  virtual bool holdTransactions(uint64_t timeout) {return _rwLock.writeLock(timeout);};
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBTransactionManager.h]: `virtual` not required here, but should be marked as `override` instead.
```suggestion
  bool holdTransactions(uint64_t timeout) override {return _rwLock.writeLock(timeout);}
```

File: arangod/RocksDBEngine/RocksDBTransactionManager.h [hunk 2]
```diff
@@ -67,8 +69,17 @@ class RocksDBTransactionManager final : public TransactionManager {
...
+  virtual bool holdTransactions(std::chrono::microseconds timeout) {return _rwLock.writeLock(timeout);};
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBTransactionManager.h]:
```suggestion
  bool holdTransactions(std::chrono::microseconds timeout) override {return _rwLock.writeLock(timeout);}
```

File: arangod/RocksDBEngine/RocksDBTransactionManager.h [hunk 3]
```diff
@@ -67,8 +69,17 @@ class RocksDBTransactionManager final : public TransactionManager {
...
+
+  // remove the block
+  virtual void releaseTransactions() {_rwLock.unlockWrite();};
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBTransactionManager.h]:
```suggestion
  void releaseTransactions() override {_rwLock.unlockWrite();}
```

File: arangod/StorageEngine/TransactionManager.h [hunk 4]
```diff
@@ -60,6 +62,14 @@ class TransactionManager {
       std::function<void(TRI_voc_tid_t, TransactionData const*)> const& callback) = 0;
 
   virtual uint64_t getActiveTransactionCount() = 0;
+
+  // functions to block transactions during hotbackup or restore
+  //  (not yet implemented/needed in all storage engines)
+  // holdTransactions: true on successful hold, false on timeout or error
+  virtual bool holdTransactions(uint64_t timeout) {return true;};
```
Comment (jsteemann) [inline arangod/StorageEngine/TransactionManager.h]:
```suggestion
  virtual bool holdTransactions(uint64_t timeout) {return true;}
```

File: arangod/StorageEngine/TransactionManager.h [hunk 5]
```diff
@@ -60,6 +62,14 @@ class TransactionManager {
...
+  virtual bool holdTransactions(std::chrono::microseconds timeout) {return true;};
```
Comment (jsteemann) [inline arangod/StorageEngine/TransactionManager.h]:
```suggestion
  virtual bool holdTransactions(std::chrono::microseconds timeout) {return true;}
```

File: arangod/StorageEngine/TransactionManager.h [hunk 6]
```diff
@@ -60,6 +62,14 @@ class TransactionManager {
...
+
+  virtual void releaseTransactions() {};
```
Comment (jsteemann) [inline arangod/StorageEngine/TransactionManager.h]:
```suggestion
  virtual void releaseTransactions() {}
```

File: lib/Basics/ReadWriteLock.cpp [hunk 7]
```diff
@@ -50,6 +50,33 @@ void ReadWriteLock::writeLock() {
   }
 }
 
+/// @brief lock for writes with microsecond timeout
+bool ReadWriteLock::writeLock(std::chrono::microseconds timeoutx) {
```
Comment (jsteemann) [inline lib/Basics/ReadWriteLock.cpp]: why is the variable named `timeoutx`? what does the `x` stand for?

Comment (matthewvon) [reply lib/Basics/ReadWriteLock.cpp]: x stands for, what the hell is keeping this function from compiling, is timeout somehow a keyword?  It was not, but I forgot to remove x.  That and all other changes are about to check in.

**Comment by matthewvon** [comment]
current hack for unregisterTransaction works:  http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/2871/  ... completed successfully

**Pull request merged**